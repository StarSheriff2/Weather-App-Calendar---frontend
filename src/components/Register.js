/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import {
  Formik, Field, Form, ErrorMessage,
} from 'formik';
import * as Yup from 'yup';

import { register } from '../slices/auth';
import { clearMessage } from '../slices/message';

const Register = () => {
  const [successful, setSuccessful] = useState(false);

  const { message } = useSelector((state) => state.message);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(clearMessage());
  }, [dispatch]);

  const initialValues = {
    name: '',
    email: '',
    password: '',
    passwordConfirmation: '',
  };

  const validationSchema = Yup.object().shape({
    name: Yup.string()
      .test(
        'len',
        'The name must be between 3 and 20 characters.',
        (val) => val && val.toString().length >= 3 && val.toString().length <= 20,
      )
      .required('This field is required!'),
    email: Yup.string()
      .test(
        'len',
        'Email is too long (maximum is 65 characters).',
        (val) => val && val.toString().length >= 3 && val.toString().length <= 65,
      )
      .email('This is not a valid email.')
      .required('This field is required!'),
    password: Yup.string()
      .test(
        'len',
        'The password must be between 6 and 40 characters.',
        (val) => val && val.toString().length >= 6 && val.toString().length <= 40,
      )
      .required('This field is required!'),
    passwordConfirmation: Yup.string()
      .oneOf([Yup.ref('password'), null], 'Passwords must match')
      .required('This field is required!'),
  });

  const handleRegister = (formValue) => {
    const {
      name, email, password, passwordConfirmation,
    } = formValue;
    setSuccessful(false);

    dispatch(register({
      name, email, password, passwordConfirmation,
    }))
      .unwrap()
      .then(() => {
        setSuccessful(true);
      })
      .catch(() => {
        setSuccessful(false);
      });
  };

  return (
    <div className="col-md-12 signup-form">
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />
        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={handleRegister}
        >
          <Form>
            {!successful && (
              <div>
                <div className="form-group">
                  <label htmlFor="name">Name</label>
                  <Field name="name" type="text" className="form-control" />
                  <ErrorMessage
                    name="name"
                    component="div"
                    className="alert alert-danger"
                  />
                </div>

                <div className="form-group">
                  <label htmlFor="email">Email</label>
                  <Field name="email" type="email" className="form-control" />
                  <ErrorMessage
                    name="email"
                    component="div"
                    className="alert alert-danger"
                  />
                </div>

                <div className="form-group">
                  <label htmlFor="password">Password</label>
                  <Field
                    name="password"
                    type="password"
                    className="form-control"
                  />
                  <ErrorMessage
                    name="password"
                    component="div"
                    className="alert alert-danger"
                  />
                </div>

                <div className="form-group">
                  <label htmlFor="passwordConfirmation">
                    Password Confirmation
                  </label>
                  <Field
                    name="passwordConfirmation"
                    type="password"
                    className="form-control"
                  />
                  <ErrorMessage
                    name="passwordConfirmation"
                    component="div"
                    className="alert alert-danger"
                  />
                </div>

                <div className="form-group">
                  <button type="submit" className="btn btn-primary btn-block">
                    Sign Up
                  </button>
                </div>
              </div>
            )}
          </Form>
        </Formik>
      </div>

      <p className="text-center mt-3">Already have an account?</p>
      <p className="text-center">
        Sign in
        {' '}
        <Link to="/">here</Link>
      </p>

      {message && (
        <div className="form-group">
          <div
            className={
              successful ? 'alert alert-success' : 'alert alert-danger'
            }
            role="alert"
          >
            {message}
          </div>
        </div>
      )}
    </div>
  );
};

export default Register;
