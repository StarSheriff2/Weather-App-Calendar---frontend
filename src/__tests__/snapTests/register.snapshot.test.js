import React from 'react';
import { cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderWithRedux from '../../tests/test-utils';
import Register from '../../components/Register';

afterEach(cleanup);

describe('Register', () => {
  test('Renders the Component', () => {
    const register = renderWithRedux(
      <Router>
        <Register />
      </Router>,
    ).toJSON();

    expect(register).toMatchSnapshot();
  });
});
