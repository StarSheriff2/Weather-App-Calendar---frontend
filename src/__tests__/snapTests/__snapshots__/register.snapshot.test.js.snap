// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Renders the Component 1`] = `
<div
  className="col-md-12 signup-form"
>
  <div
    className="card card-container"
  >
    <img
      alt="profile-img"
      className="profile-img-card"
      src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
    />
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div>
        <div
          className="form-group"
        >
          <label
            htmlFor="name"
          >
            Name
          </label>
          <input
            className="form-control"
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="email"
          >
            Email
          </label>
          <input
            className="form-control"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="form-control"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="passwordConfirmation"
          >
            Password Confirmation
          </label>
          <input
            className="form-control"
            name="passwordConfirmation"
            onBlur={[Function]}
            onChange={[Function]}
            type="passwordConfirmation"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <button
            className="btn btn-primary btn-block"
            type="submit"
          >
            Sign Up
          </button>
        </div>
      </div>
    </form>
  </div>
  <p
    className="text-center mt-3"
  >
    Already have an account?
  </p>
  <p
    className="text-center"
  >
    Sign in
     
    <a
      href="/"
      onClick={[Function]}
    >
      here
    </a>
  </p>
</div>
`;
